File: tests/catch2_tests/test_write_png.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2022-2026 Stephan Jaeckel (http://quadriga-lib.org)
// Part of quadriga-lib — see LICENSE for terms.

#include <catch2/catch_test_macros.hpp>

#include "quadriga_tools.hpp"

TEST_CASE("Write PNG")
{
    const arma::uword N = 500;  // size along one edge
    const double maxVal = 10.0; // value at the upper‑right corner
    arma::mat gradient(N, N, arma::fill::none);
    const double scale = maxVal / double(N); // (= 10 / 98 for N = 50)

    for (arma::uword r = 0; r < N; ++r)
        for (arma::uword c = 0; c < N; ++c)
            gradient(r, c) = scale * c;

    quadriga_lib::write_png(gradient, "test.png");

    REQUIRE(std::filesystem::exists("test.png"));

    std::remove("test.png");
}

File: tests/python_tests/test_write_png.py
import sys
import os
import unittest
import numpy as np
import numpy.testing as npt

# Append the directory containing your package to sys.path
current_dir = os.path.dirname(os.path.abspath(__file__))
package_path = os.path.join(current_dir, '../../lib')
if package_path not in sys.path:
    sys.path.append(package_path)

import quadriga_lib

class test_case(unittest.TestCase):

    def test(self):

        N = 50                 # size along one edge
        max_val = 10.0          # value at the upper‑right corner

        # Create a left‑to‑right gradient (columns vary, rows repeat)
        gradient = np.tile(np.linspace(0.0, max_val, N, endpoint=False), (25, 1))
        gradient[0,:] = 0.0
        gradient[-1,:] = 10.0

        # File name to write
        out_file = "test.png"

        # quadriga_lib.tools.write_png(fn, data, colormap, min_val, max_val, log_transform)
        quadriga_lib.tools.write_png(
            out_file,
            gradient,
            "jet",          # any valid Matplotlib‑style colormap string works
            0.0,                # min_val  (matches gradient.min())
            max_val,            # max_val  (matches gradient.max())
            False               # log_transform
        )

        # Check that the file was created
        self.assertTrue(os.path.exists(out_file))

        # Clean up
        os.remove(out_file)